src/schedule/zephyr_preempt_domain.c

```diff
@@ -0,0 +1,460 @@
+	struct k_sem sem; /* used to signal when work should be done */
+	void (*handler)(void *arg); /* work to be done */
+	void *arg; /* data used by work function */
+	bool started; /* true if the thread was started */
+};
+
+struct zephyr_preempt_domain {
+	struct dma *dma_array; /* array of scheduling DMAs */
+	uint32_t num_dma; /* number of scheduling DMAs */
+
+	struct zephyr_preempt_domain_data data[PLATFORM_NUM_DMACS][PLATFORM_MAX_DMA_CHAN];
+	/* array of Zephyr threads - one for each core */
+	struct zephyr_preempt_domain_thread domain_thread[CONFIG_CORE_COUNT];
+};
+
+static int zephyr_preempt_domain_register(struct ll_schedule_domain *domain,
+					  struct task *task,
+					  void (*handler)(void *arg),
+					  void *arg);
+static int zephyr_preempt_domain_unregister(struct ll_schedule_domain *domain,
+					    struct task *task,
+					    uint32_t num_tasks);
+static void zephyr_preempt_domain_task_cancel(struct ll_schedule_domain *domain,
+					      uint32_t num_tasks);
+
+static const struct ll_schedule_domain_ops zephyr_preempt_domain_ops = {
+	.domain_register	= zephyr_preempt_domain_register,
+	.domain_unregister	= zephyr_preempt_domain_unregister,
+	.domain_task_cancel	= zephyr_preempt_domain_task_cancel
+};
+
+struct ll_schedule_domain *zephyr_preempt_domain_init(struct dma *dma_array,
+						      uint32_t num_dma,
+						      int clk)
+{
+	struct ll_schedule_domain *domain;
+	struct zephyr_preempt_domain *zephyr_preempt_domain;
+
+	/* initialize domain */
+	domain = domain_init(SOF_SCHEDULE_LL_DMA,
+			     clk,
+			     true,
+			     &zephyr_preempt_domain_ops);
+
+	/* initialize domain pdata */
+	zephyr_preempt_domain = rzalloc(SOF_MEM_ZONE_SYS_SHARED,
+					0,
+					SOF_MEM_CAPS_RAM,
+					sizeof(*zephyr_preempt_domain));
+
+	zephyr_preempt_domain->dma_array = dma_array;
+	zephyr_preempt_domain->num_dma = num_dma;
+
+	/* set pdata */
+	ll_sch_domain_set_pdata(domain, zephyr_preempt_domain);
+
+	return domain;
+}
+
+static void zephyr_preempt_domain_thread_fn(void *p1, void *p2, void *p3)
+{
+	struct zephyr_preempt_domain_thread *dt = p1;
+
+	while (true) {
+		/* wait for DMA IRQ */
+		k_sem_take(&dt->sem, K_FOREVER);
+
+		/* do work */
+		dt->handler(dt->arg);
+	}
+}
+
+static void dma_irq_handler(void *data)
+{
+	struct zephyr_preempt_domain_data *zephyr_preempt_domain_data;
+	struct zephyr_preempt_domain_thread *dt;
+	struct dma_chan_data *channel;
+	struct k_sem *sem;
+	int channel_index, irq;
+
+	zephyr_preempt_domain_data = data;
+	channel = zephyr_preempt_domain_data->channel;
+	channel_index = channel->index;
+	sem = &zephyr_preempt_domain_data->dt->sem;
+	irq = zephyr_preempt_domain_data->irq;
+	dt = zephyr_preempt_domain_data->dt;
+
+	/* give resources to thread semaphore */
+	if (dt->handler)
+		k_sem_give(sem);
+
+	/* clear IRQ */
+	dma_interrupt_legacy(channel, DMA_IRQ_CLEAR);
+	interrupt_clear_mask(irq, BIT(channel_index));
+}
+
+static int enable_dma_irq(struct zephyr_preempt_domain_data *data)
```

Comment:
this could be timer, can we rename ```_dma_``` to ```_prempt_```

Reply:
unless this code is dma specific ? and we need a similar function to pass in another wake source

Reply:
The only similarity between let's say a timer enable function and the DMA one is the fact that we need to register a handler and then enable the IRQ. For now, I say we leave this as is so as to avoid any confusions and we can change it later on when the support for timer IRQs will be added.